compiler/src/iree/compiler/Codegen/Common/TileDispatchUsingForall.cpp

```diff
@@ -0,0 +1,344 @@
+///
+/// ```
+/// [workgroup_mapping<x>, workgroup_mapping<y>,
+///  workgroup_mapping<z:1>, workgroup_mapping<z>]
+/// ```
+SmallVector<Attribute>
+pruneDroppedLoops(ArrayRef<Attribute> inputs,
+                  const llvm::SmallDenseSet<int> &droppedLoops) {
+  SmallVector<IREE::Codegen::WorkgroupMappingAttr> droppedMappings;
+  SmallVector<Attribute> prunedAttrs;
+  for (auto [index, input] : llvm::enumerate(inputs)) {
+    if (droppedLoops.contains(index)) {
+      droppedMappings.push_back(
+          cast<IREE::Codegen::WorkgroupMappingAttr>(input));
+    } else {
+      prunedAttrs.push_back(input);
+    }
+  }
+  for (auto droppedMapping : droppedMappings) {
+    for (auto [index, prunedAttr] : llvm::enumerate(prunedAttrs)) {
+      auto prunedMappingAttr =
+          cast<IREE::Codegen::WorkgroupMappingAttr>(prunedAttr);
+      if (droppedMapping < prunedMappingAttr) {
+        prunedAttrs[index] =
+            IREE::Codegen::WorkgroupMappingAttr::getAttributeFromMappingId(
+                prunedAttr.getContext(), prunedMappingAttr.getMappingId() - 1);
+      }
+    }
+  }
+  return prunedAttrs;
+}
+
+// Find dimensions of the loop that are unit-trip count and drop them from the
+// distributed dimensions.
+struct DropUnitDistributedDims final : public OpRewritePattern<scf::ForallOp> {
+  using OpRewritePattern<scf::ForallOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(scf::ForallOp forallOp,
+                                PatternRewriter &rewriter) const override {
+    SmallVector<OpFoldResult> mixedLbs = forallOp.getMixedLowerBound();
+    SmallVector<OpFoldResult> mixedUbs = forallOp.getMixedUpperBound();
+    SmallVector<OpFoldResult> mixedSteps = forallOp.getMixedStep();
+
+    // Find the index of loops to be dropped.
+    llvm::SmallDenseSet<int> droppedLoops;
+    for (auto [index, lb, ub, step] :
+         llvm::enumerate(mixedLbs, mixedUbs, mixedSteps)) {
+      if (!isa<Attribute>(lb) || !isa<Attribute>(ub) || !isa<Attribute>(step)) {
+        continue;
+      }
+      int64_t lbVal = getConstantIntValue(lb).value();
+      int64_t ubVal = getConstantIntValue(ub).value();
+      int64_t stepVal = getConstantIntValue(step).value();
+      if (CEILDIV(ubVal - lbVal, stepVal) == 1) {
+        droppedLoops.insert(index);
+      }
+    }
+    if (droppedLoops.empty()) {
+      return rewriter.notifyMatchFailure(forallOp, "no loops to drop");
+    }
+
+    SmallVector<OpFoldResult> newLbs =
+        pruneDroppedLoops(ArrayRef<OpFoldResult>(mixedLbs), droppedLoops);
+    SmallVector<OpFoldResult> newUbs =
+        pruneDroppedLoops(ArrayRef<OpFoldResult>(mixedUbs), droppedLoops);
+    SmallVector<OpFoldResult> newSteps =
+        pruneDroppedLoops(ArrayRef<OpFoldResult>(mixedSteps), droppedLoops);
+    std::optional<ArrayAttr> newMapping;
+    if (auto currMapping = forallOp.getMapping()) {
+      SmallVector<Attribute> newMappingAttrs =
+          pruneDroppedLoops(currMapping.value().getValue(), droppedLoops);
```

Comment:
The casts in `pruneDroppedLoops` will cause crashes if this pass is run with a non-workgroup forall present. Can you add a check for `WorkgroupMappingAttr::verifyAttrList` or something similar here.

Reply:
The list is verified already when the mapping is computed before tiling. Its the same mapping that is added to the loop. But you are right, we dont want to run this always. Changed it to only run on the loop generated from tiling.